xmtp_mls/src/groups/device_sync.rs

```diff
@@ -0,0 +1,596 @@
+        &self,
+        kind: DeviceSyncKind,
+    ) -> Result<(StoredGroupMessage, DeviceSyncRequestProto), DeviceSyncError> {
+        let sync_group = self.get_sync_group()?;
+
+        sync_group.sync().await?;
+
+        let messages = sync_group
+            .find_messages(&MsgQueryArgs::default().kind(GroupMessageKind::Application))?;
+
+        for msg in messages.into_iter().rev() {
+            let msg_content: DeviceSyncContent =
+                serde_json::from_slice(&msg.decrypted_message_bytes)?;
+            match msg_content {
+                DeviceSyncContent::Reply(reply) if reply.kind() == kind => {
+                    return Err(DeviceSyncError::NoPendingRequest);
+                }
+                DeviceSyncContent::Request(request) if request.kind() == kind => {
+                    return Ok((msg, request));
+                }
+                _ => {}
+            }
+        }
+
+        Err(DeviceSyncError::NoPendingRequest)
+    }
+
+    /// Look for sync reply by kind, returns NoReplyToProcess error if not found.
+    async fn sync_reply(
+        &self,
+        kind: DeviceSyncKind,
+    ) -> Result<DeviceSyncReplyProto, DeviceSyncError> {
+        let sync_group = self.get_sync_group()?;
+
+        sync_group.sync().await?;
+        let messages = sync_group
+            .find_messages(&MsgQueryArgs::default().kind(GroupMessageKind::Application))?;
+
+        for msg in messages.iter().rev() {
+            // ignore this installation's messages
+            if msg.sender_installation_id == self.installation_public_key() {
+                continue;
+            }
+
+            let content: DeviceSyncContent = serde_json::from_slice(&msg.decrypted_message_bytes)?;
+            if let DeviceSyncContent::Reply(reply) = content {
+                if reply.kind() == kind {
+                    return Ok(reply);
+                }
+            }
+        }
+
+        Err(DeviceSyncError::NoReplyToProcess)
+    }
+
+    async fn process_sync_reply(&self, kind: DeviceSyncKind) -> Result<(), DeviceSyncError> {
+        let conn = self.store().conn()?;
+
+        let reply = self.sync_reply(kind).await?;
+
+        let time_diff = reply.timestamp_ns.abs_diff(now_ns() as u64);
+        if time_diff > NS_IN_HOUR as u64 {
+            // time discrepancy is too much
+            return Err(DeviceSyncError::SyncReplyTimestamp);
+        }
+
+        let Some(enc_key) = reply.encryption_key.clone() else {
+            return Err(DeviceSyncError::InvalidPayload);
+        };
+
+        let enc_payload = download_history_payload(&reply.url).await?;
+        insert_encrypted_syncables(&conn, enc_payload, &enc_key.try_into()?)?;
+
+        self.sync_welcomes().await?;
+
+        let groups = conn.find_groups(None, None, None, None, Some(ConversationType::Group))?;
+        for crate::storage::group::StoredGroup { id, .. } in groups.into_iter() {
+            let group = self.group(id)?;
+            Box::pin(group.sync()).await?;
```

Comment:
related to connection use, wherever we do a `sync`, we should probably actually be doing a `sync_with_conn` in order to re-use the connection. That requires the `XmtpOpenMlsProvider` which can be created from an owned value of `DbConnection`.

Anything not explicitly requiring the provider (like normal queries) can still access the connection with `conn_ref`